### PR TITLE
it will delete old page while adding new page 

### LINKED ISSUE #1984: Renaming a page doesn't delete old page
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

1. Update name of any page
2. Then see if the changes remain preserved after you restart the app.

Also, in the preview the new name doesn't show.

### Current behavior 😯

https://user-images.githubusercontent.com/92228082/237060570-34f55fb0-7ac5-4597-8e3c-07cf22632d86.mov

### Expected behavior 🤔

_No response_

### Context 🔦

_No response_

### Your environment 🌎

Latest (v0.1.10)

### PR DESCRIPTION
**Pull request by JerryWu1234**
fix #1984
it will delete old page while adding new page 
- [x] I've read and followed the [contributing guide](https://github.com/mui/mui-toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [x] I've updated the relevant documentation for any new or updated feature
- [x] I've linked relevant GitHub issue with "Closes #<issue id>"
- [x] I've added a visual demonstration under the form of a screenshot or video

### CODE REVIEW

**Comment by JerryWu1234** [comment]
image

**Comment by Janpot** [comment]
- Would it be possible to do this similarly to `deletePage` and just rename the page folder on the file system?
- This needs a test

**Comment by JerryWu1234** [comment]
> * Would it be possible to do this similarly to `deletePage` and just rename the page folder on the file system?

I totally agree, and I will follow your idea.
I will create a new function that calls `renamefolder`.

Could I ask you a question?
actually, I haven't found out where it tests part of the `deletePage` in project
please point out if you are available

**Comment by JerryWu1234** [comment]
image
image


@Janpot sorry I'm not good at playwright but I tried a lot of methods.
unfortunately, I can't run the unit test in my local. I don't know why,
please help if you are available.

image

**Comment by Janpot** [comment]
I've added a guide on how to run integration tests locally. Please take a look and let me know whether this helps you further: https://github.com/mui/mui-toolpad/pull/2192

**Comment by JerryWu1234** [comment]
> I've added a guide on how to run integration tests locally. Please take a look and let me know whether this helps you further: #2192

I'm sorry for replying late.
I'll try late

**Comment by JerryWu1234** [comment]
> I've added a guide on how to run integration tests locally. Please take a look and let me know whether this helps you further: #2192

image
Some parts of the unit test will not succeed  but some part not.
I used both methods the dev mode and production mode.
Still, I got the same result, although I followed your steps.

**Comment by JerryWu1234** [comment]
@Janpot 
image

**Comment by JerryWu1234** [comment]
> > parts of the unit test will not succeed but some part not. I used both methods the dev mode and production mode. Still, I got the same result, although I followed your steps.
> 
> You are right, I messed up the environment variable. The correct command is:
> 
> ```shell
> TOOLPAD_NEXT_DEV=1 yarn test:integration --project chromium
> ```
> 
> #2199

image

Better now. But it has also some errors in it.

**Comment by Janpot** [comment]
I works for me this way. 
* Which assertions are failing? 
* What does the playwright console show? 

  Screenshot 2023-06-22 at 10 53 48

* Does it work when you're on master?

**Comment by JerryWu1234** [comment]
> 

@Janpot 
image
this is the log.
No matter which branch you choose, the result is the same.
looks like it doesn't access that website

**Comment by Janpot** [comment]
I've seen this happen some time on the first run. Try whether tit works by running the integration tests in production mode.
I'll increase the timeout for this use case in https://github.com/mui/mui-toolpad/pull/2222

**Comment by JerryWu1234** [comment]
thanks.
 I'll try it when it is merged

**Comment by JerryWu1234** [comment]
> I've seen this happen some time on the first run. Try to verify whether it works by running the integration tests in production mode. I'll increase the timeout for this use case in #2222

it's fine for Pro，
image

But it is still wrong in dev 
image

**Comment by JerryWu1234** [comment]
@Janpot I added the unit test.

**Comment by JerryWu1234** [comment]
> > I've seen this happen some time on the first run. Try to verify whether it works by running the integration tests in production mode. I'll increase the timeout for this use case in #2222
> 
> @Janpot it's fine for Pro， image
> 
> But it is still wrong in dev image

@Janpot 
Waiting for a while, everything will be better if you run yarn run dev
But tests can pass sometimes, sometimes not.
However, at least it can’t take effect for writing test.

**Comment by JerryWu1234** [comment]
thanks again

File: packages/toolpad-app/src/server/localMode.ts [hunk 1]
```diff
@@ -760,9 +766,17 @@ async function writeThemeFile(root: string, theme: Theme | null) {
   }
 }
 
-async function writeDomToDisk(dom: appDom.AppDom): Promise<void> {
+async function writeDomToDisk(dom: appDom.AppDom, oldDom?: appDom.AppDom): Promise<void> {
```
Comment (Janpot) [inline packages/toolpad-app/src/server/localMode.ts:769]: Is it be possible to do this the way `deletePage` works? Where we call `deletePage` from the frontend and don't rely on `writeDomToDisk` to reconcile everything? 

The path forward for the `AppDom` will probably be to split it up in its separate pieces and not read/write it as a whole.

**Comment by JerryWu1234** [comment]
image

@Janpot 
are There only four people? LOL

**Comment by prakhargupta1** [comment]
@JerryWu1234 We started the Discord server last week and haven't shared it with anyone yet.

**Comment by JerryWu1234** [comment]
> @JerryWu1234 We started the Discord server last week and haven't shared it with anyone yet.

I'm just curious about whether you will discuss all their contributed problems on the Discord server in the future?

**Comment by JerryWu1234** [comment]
@Janpot I done。help check

File: test/integration/editor/new.spec.ts [hunk 2]
```diff
@@ -77,3 +76,20 @@ test('can create/delete page', async ({ page, localApp }) => {
...
 });
+
+test('can rename page', async ({ page, localApp }) => {
+  const editorModel = new ToolpadEditor(page);
+
+  await editorModel.goto();
+
+  await editorModel.createPage('someOtherPage');
+
+  const pageFolder = path.resolve(localApp.dir, './toolpad/pages/someOtherPage');
```
Comment (Janpot) [inline test/integration/editor/new.spec.ts:87]: Just to protect ourselves from accidentally breaking this test, let's double verify the folder exists
```suggestion
  const pageFolder = path.resolve(localApp.dir, './toolpad/pages/someOtherPage');
  await expect.poll(async () => folderExists(pageFolder)).toBe(true);
```

File: packages/toolpad-app/src/toolpad/AppEditor/NodeNameEditor.tsx [hunk 3]
```diff
@@ -25,13 +26,19 @@ export default function NodeNameEditor({ node, sx }: NodeNameEditorProps) {
...
-      domApi.setNodeName(node.id, nameInput);
+      if (nameInput !== oldname.name) {
+        domApi.setNodeName(node.id, nameInput);
+      }
     } else {
       setNameInput(node.name);
     }
-  }, [isNameValid, domApi, node.id, node.name, nameInput]);
+    if (nameInput !== oldname.name) {
+      await client.mutation.deletePage(oldname.name);
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/AppEditor/NodeNameEditor.tsx:39]: `NodeNameEditor` is used in several other places than pages. e.g. https://github.com/mui/mui-toolpad/blob/564108a247a30b6a570974308732020a6e1d479b/packages/toolpad-app/src/toolpad/AppEditor/PageEditor/ComponentEditor.tsx#L175

Wouldn't this break other usage of this component? I can't imagine a solution where this component would need to import 
```tsx
import client from '../../api';
```

**Comment by JerryWu1234** [comment]
@Janpot I added a test for what you said.
help check thanks

**Comment by JerryWu1234** [comment]
I added a test that compares the previous node with the new node created, to test `NodeNameEditor.tsx`。
hopefully, I get it right

**Comment by JerryWu1234** [comment]
image

**Comment by Janpot** [comment]
Ok, we have integration tests, we can always improve the implementation later. I've added the following updates:

* add test for when the folder is renamed on disk
* fix for when folder is renamed on disk (`chokidar` doesn't seem to detect this unless polling is used)

Thanks for this fix, we can merge soon

**Review by Janpot — APPROVED** [review]

**Pull request merged**